### DISCUSSION TITLE
proposal for allowing plugins to expose slot props inside forms

### DISCUSSION THREAD

**Discussion by logaretm** [formvuelate/formvuelate#180] [Ideas]
# What

This is a proposal that explores a way to allows plugins to expose slot props on the `SchemaForm` component `beforeForm` and `afterForm` slots.

The plugins that "add state" to formvuelate will need somehow to expose their state to the user, either via template or composition API. This is [especially clear here in this issue](https://github.com/formvuelate/formvuelate-plugin-vee-validate/issues/8)

Now, the workaround the author suggested would work fine but they do make use of `$refs` which I'm not a fan of due to reactivity issues, although Vue 3 does offer ways to make it reactive.

# How

I'm suggesting implementing a `slotProps` property that can be injected in the schema that will be bound to the `beforeForm` and `afterForm` slots. 

Here is a couple of usage examples:

```js
// inside the plugin fn
// ...
const formSlotProps = computed(() => {
  return {
    validation: {
      meta: meta.value,
      errors: errors.value,
      submitCount: submitCount.value,
      isSubmitting: isSubmitting.value,
      resetForm,
      setTouched,
    }
  }
});

return {
  ...baseReturns,
  formBinds: computed(() => {
    return {
      ...baseReturns.formBinds.value,
      // maybe place those in a different place, like on the root schema but that's trivial
      // also maybe both slots should have the same props which would eliminate having 2 properties for slots
      slotProps: formSlotProps.value,
    }
  }),

  parsedSchema: formSchemaWithVeeValidate
}
```

This could allow us to do common things, in vee-validate's case it would open up these possibilities:

**Disable buttons until all fields are valid**

```vue
<SchemaWithValidation :schema="schema">
  <template #afterForm="{ validation }">
    <button :disabled="!validation.meta.valid">Submit</button>
  </template>
</SchemaWithValidation>
```

**Display spinner or loading state when the form is submitting**

```vue
<SchemaWithValidation :schema="schema">
  <template #afterForm="{ validation }">
    <button :class="{ 'is-submitting': validation.isSubmitting }">Submit</button>
  </template>
</SchemaWithValidation>
```

**Disable submit until the user makes changes to form values**

```vue
<SchemaWithValidation :schema="schema">
  <template #afterForm="{ validation }">
    <button :disabled="!validation.meta.dirty">Submit</button>
  </template>
</SchemaWithValidation>
```

**Display all errors as a summary before the form fields**

```vue
<SchemaWithValidation :schema="schema">
  <template #beforeForm="{ validation }">
    <p>Please correct these errors</p>
    <ul v-if="validation.errors">
      <li v-for="error in validation.errors">{{ error }}</li>
    </ul>
  </template>
</SchemaWithValidation>
```

The downside of this approach is that if the developer wanted to access these extended properties/props inside their JS block, it will be very tricky. This might suggest adding a `useXXXXXX` function to complement this API but that's a different topic and may need its own proposal.

**Comment by marina-mosti** [comment]
Awesome thank you for reposting it here @logaretm
